# Get a list of BAM files matching the pattern "deduplicated*.bam"
bam_files <- list.files(pattern = "deduplicated.*\\.bam$")

# Define the threshold value
thresholds <- c(0.0001)

# Create an empty data frame to store the  results
results <- data.frame()

for (threshold in thresholds) {
  # Define the output prefix for each threshold
  output_prefix <- paste0("output_ivar_table_", threshold)

  # Get a list of BAM files matching the pattern "deduplicated*.bam"
  bam_files <- list.files(pattern = "deduplicated.*\\.bam$")

  # Initialize an empty list to store the results for each BAM file
  results_list <- list()
  
    # Iterate through each BAM file
  
      for (bam_file in bam_files) {
      # Execute the ivar variants command with the current threshold and BAM file
        command <- paste("/opt/homebrew/bin/samtools mpileup -aa -A -d 0 -B -Q 20", bam_file, " |/opt/homebrew/bin/ivar variants -p", output_prefix, "-q 20 -t", threshold, "-r CHIKV_genome.fa -g CHIKV_genome.gff")
      system(command)
      
      # Read the TSV file generated by ivar variants
      tsv_file <- paste0(output_prefix, ".tsv")
      tsv_data <- read.table(tsv_file, header = TRUE, sep = "\t")
      
      # Add a column for the threshold value
      tsv_data$Threshold <- threshold
      tsv_data$bam_file <- bam_file
      tsv_data$condition <- str_extract(tsv_data$bam_file, "MFM[^_]*-[^_]*-[^_]*-[^_]*_S\\d+")
      
      # Append the results to the combined data frame
      results <- rbind(results, tsv_data)
      }
}
#Remove rows with less than 10 TOT_DP
combined_results <- combined_results[which(combined_results$TOTAL_DP > 500), ]
#Remove rows with PASS=FALSE
combined_results <- combined_results[which(combined_results$PASS == "TRUE"), ]

#Recalculate the TOTAL_DP
combined_results_0 <- data_frame()

for (bam_file in bam_files){
  combined_results_1 <-  combined_results[which(combined_results$bam_file == bam_file), ]
  
  for (threshold in thresholds) {
    # Filter the tsv dataframe based on the current threshold
    combined_results_2 <- combined_results_1[which(combined_results_1$Threshold == threshold), ]
    
    variant_DP <- combined_results_2 %>%
      group_by(POS) %>%
      mutate(
        total_DP_variant =  sum(ALT_DP), # Replace calc_variant_logic with your logic
      ) %>%
      ungroup()
    
    variant_DP$TOTAL_DP <- variant_DP$REF_DP + variant_DP$total_DP_variant
    combined_results_0 <- rbind(combined_results_0, variant_DP)
    }
}

#Remove variants with less than depth = 500
results <- results[which(results$TOTAL_DP > 500), ]
#Recalculate the ALT_FREQ
results$ALT_FREQ = 1-(results$TOTAL_DP-results$ALT_DP)/results$TOTAL_DP
##check variants for strand biais
# Calculate ALT_FW
results$ALT_FW <- results$ALT_DP - results$ALT_RV
results$REF_FW <- results$REF_DP - results$REF_RV

# Perform Fisher's exact test on the calculated values
results$Strand_bias <- as.logical("")

fisher_results <- for (i in 1:nrow(combined_results)) {
  counts <- results[i, c("REF_FW", "REF_RV", "ALT_FW", "ALT_RV")]
  table <- matrix(as.numeric(counts), nrow = 2)
  test_result <- fisher.test(table, alternative = "greater")
  results[i,26] <- test_result$p.value < 0.05
}

#if TRUE, there is a strand bias and the variant should be rejected
results_filtered <- results[which(results$Strand_bias == "FALSE"),]

##get AA number 
nsp4_start <- 5666  # Start position for CHIKV_NSP4
nsp1_3_start <- 77   # Start position for CHIKV_NSP1_3
structural_start <- 7567 # Start position for CHIK_STRUCTURAL

results_filtered$AA_num <- sapply(1:nrow(results_filtered), function(i) {
  feature <- results_filtered$GFF_FEATURE[i]
  pos <- results_filtered$POS[i]
  
  if (!is.na(feature)) {
    if (feature == "CHIKV_NSP1_3") {
      return(ceiling((pos - (nsp1_3_start - 1)) / 3))
    } else if (feature == "CHIKV_NSP4") {
      return(ceiling((pos - (nsp4_start - 1)) / 3))
    } else if (feature == "CHIK_STRUCTURAL") {
      return(ceiling((pos - (structural_start - 1)) / 3))
    } else {
      return(NA)  # Default case if the feature does not match any condition
    }
  } else {
    return(NA)  # Case for NA in GFF_FEATURE
  }
})

#add AA mutation
results_filtered$AA_mutation <- paste(combined_results_filtered$REF_AA, combined_results_filtered$AA_num, combined_results_filtered$ALT_AA)

#write table
write.table(results_filtered, file = "Variant_caller_variants_no_strand_bias.txt", sep = "\t", row.names = FALSE)
